hal/src/photon/concurrent_hal.cpp

```diff
@@ -114,11 +103,239 @@ os_result_t os_thread_join(os_thread_t* thread)
  * @param thread    The thread to clean up.
  * @return 0 on success.
  */
-os_result_t os_thread_cleanup(os_thread_t* thread)
+os_result_t os_thread_cleanup(os_thread_t thread)
 {
-    return wiced_rtos_delete_thread((wiced_thread_t*)thread);
+    vTaskDelete(thread);
+    return 0;
 }
 
+/**
+ * Map gthread handles to FreeRTOS task handles.
+ */
+typedef TaskHandle_t __gthread_t;
+
+
+class ThreadQueue
+{
+    QueueHandle_t queue;
+
+public:
+
+    ThreadQueue(UBaseType_t max_size) {
+        queue = xQueueCreate(max_size, sizeof(TaskHandle_t));
+    }
+
+    ~ThreadQueue() {
+        vQueueDelete(queue);
+    }
+
+
+    void enqueue(TicksType_t ticks=portMAX_DELAY) {
+        TaskHandle_t current = xTaskGetCurrentTaskHandle();
+        xQueueSend(queue, &current, ticks);
+    }
+
+    void sleep() {
+        enqueue();
+        vTaskSuspend(NULL);
+    }
+
+    bool wake() {
+        TaskHandle_t next;
+        if (xQueueReceive(queue, &next, 0)==pdTRUE) {
+            vTaskResume(next);
+            return true;
+        }
+        return false;
+    }
+
+    void wakeAll()
+    {
+        while (wake()) {}
+    }
+};
+
+
+class Lock
+{
+    SemaphoreHandle_t semaphore;
+
+public:
+
+    Lock() {
+        semaphore = xSemaphoreCreateMutex();
+    }
+
+    void acquire() {
+        while (xSemaphoreTake(semaphore, portMAX_DELAY)==pdFALSE);
+    }
+
+    void release() {
+        xSemaphoreGive(semaphore);
+    }
+
+    ~Lock() {
+        vSemaphoreDelete(semaphore);
+    }
+};
+
+
+class ConditionVariable
+{
+    ThreadQueue queue;
+
+    typedef std::unique_lock<std::mutex> lock_t;
+public:
+    ConditionVariable(UBaseType_t max_size) : queue(max_size) {}
+
+    void wait(lock_t* lock)
+    {
+        taskENTER_CRITICAL();
+        lock->unlock();
+        queue.enqueue();
+        taskEXIT_CRITICAL();
+        vTaskSuspend(NULL);
+        lock->lock();
+    }
```

Comment:
I agree that only only one task may enter wait or notify because of the mutex / shared lock. My point was that since the lock is unlocked in wait there is a possibility (however unlikely) that the other task will enter notify before the first task suspends itself.

Doing something like vTaskSuspend outside a critical section and without a lock while another thread is trying do vTaskResume around the same time has to have a failure mode.

I wrote that scenario in code. I used os_thread_scheduling to force the scheduler not to preempt the application task until right after the end of the critical section in wait, then forced the scheduler to preempt the application task (this could happen "in the wild" without the extra code). In this case vTaskResume is called before vTaskSuspend and the application task freezes forever (loop is never called in application.cpp).

https://github.com/munirent/particle-firmware/pull/1/files